text/1001-websocket-protocol.rst

```diff
@@ -0,0 +1,139 @@
+..
+    Status: Draft
+    Type: Feature
+    Created: 2020-02-19
+    RFC PR: `edgedb/rfcs#4 <https://github.com/edgedb/rfcs/pull/4>`_
+
+======================================
+RFC 0001: The WebSocket-based Protocol
+======================================
+
+This RFC describes the design and implementation of the alternate
+protocol to the EdgeDB based on WebSockets for framing and the initial
+handshake.
+
+
+Motivation
+==========
+
+The primary motivation is to allow various browser-based exploration
+and editing tools for the database.
+
+The secondary motivation is to allow browser-based access to the
+application data. Essentially allowing serverless implementation for
+some applications or parts thereof. The reason this is a secondary
+motivation is that it requires full featured access controls (ACLs)
+which we don't have yet.
+
+Two important side effects of implementing this RFC:
+* it might also help implementing the protocol in some languages which
+  we don't have language bindings yet
+* existing tools like load-balancers can be used to access edgedb in
+  some cases
+* it enables using TLS which we don't have for the main protocol yet
+
+
+Specification
+=============
+
+Changes from the Binary Protocol
+--------------------------------
+
+1. The URL to connect to is ``ws://<host>:<port>/ws/<database-name>``
```

Comment:
cc @elprans 

`/ws/<database-name>` is actually an interesting idea. We should probably require it for all our Ports. 

This way you can configure an HTTP+EDGEQL port 8888 for database 'A' (use `:8888/edgeql/A`), an HTTP+GRAPHQL port for database 'B' (use `:8888/graphql/B`) and a WS port for 'A' (use `:8888/ws/A/`).

Currently our ports listen on `/` which means that you need to use haproxy or nginx to enable several edgedb application ports on one network port.

Reply:
So are we making the `database` property on a `Port` a `MULTI` property?

Reply:
I thought we can have multiple application port objects (dropping the exclusive constraint on the network port), e.g.

```
INSERT Port { port := 8888, database := 'A', protocol := 'http+edgeql', ...};
INSERT Port { port := 8888, database := 'B', protocol := 'http+graphql', ...};
INSERT Port { port := 8888, database := 'A', protocol := 'ws+edgeql', ...};
```

Essentially a new exclusive constraint would be on `(port, database, protocol)`.

Reply:
Then calling them "ports" is not making a whole lot of sense anymore.  In my opinion, we need to take a step back and maybe rethink the design of the whole concept.

Remember, the original motivation for these is to provide a simple interface for pre-allocated connection pools.  In that sense, the only important thing in the specification is the database name and the size of the pool.  The protocol and other things are secondary and can actually reuse the same connection pool.

Reply:
> In my opinion, we need to take a step back and maybe rethink the design of the whole concept.

Yes. The current design precludes one from easily exposing both EdgeQL and GraphQL on one HTTP port which can be troubling in some network setups. Ideally there should be a way to expose whatever you want related to HTTP on one network port.

Perhaps we can still call them "Ports" just redesign the schema:

```
type Port {
  property port -> int;
  link interfaces -> Interface;
}

type Interface {
   database
   concurrency
   protocol
}
```

(I don't like the `Interface` name, so it's just a placeholder).

I'm also wondering if we should change our binary protocol to send an `\x00` byte when a socket connection is established; that would allow us to multiplex edgedb-binary protocol along with WS/HTTP on a single network port (as Paul once suggested).

Reply:
> I'm also wondering if we should change our binary protocol to send an \x00 byte when a socket connection is established; that would allow us to multiplex edgedb-binary protocol along with WS/HTTP on a single network port (as Paul once suggested).

We can already distinguish, because our protocol always starts with `V\x00`. We only need to constrain handshake to < 16Mb max to make that always true, which is sensible I think. Such connection negotiation is not okay to expose to the internet without TLS, though. But it's okay to use inside the private network.

> Then calling them "ports" is not making a whole lot of sense anymore. In my opinion, we need to take a step back and maybe rethink the design of the whole concept.

Yes, we have already discussed that connection pools and ports should be different features, allowing arbitrary mapping between ports and connection pools. But I think this is out of scope of this RFC.

----

Other than that I didn't work on configuring ports yet.

Reply:
Good point re `V\x00`.